[ENH] Kolmogorov Arnold Block for NBeats

Sohaib-Ahmed21 opened the pull request:
### Description
Fixes: #1741 

This PR adds Kolmogorov Arnold(KAN) Blocks in NBeats and also does refactoring of NBeats. Implementation of KAN blocks' layers is taken from [original paper code](https://github.com/KindXiaoming/pykan/tree/master). 

### Changes in Structure

* Introduced the `NBEATSKAN` module, which enables usage of **KAN blocks** within the `NBEATS` architecture.

* Integrated `KANLayer` logic, implemented in `kan_layer.py`, which handles KAN-specific operations such as:

  * **Spline coefficient computation**,
  * **Grid initialization and updates**, etc.

* Imported `KANLayer` to `submodules.py` for block operations, allowing `NBEATSKAN` to delegate block-level behavior through `use_kan=True`.

* Added the `NBEATSAdapter` class to encapsulate **common methods** shared by both `NBEATS` and `NBEATSKAN`, including:

  * Standard training, forward logic, etc.
  * **Excludes** block initialization (`__init__`), which is separately defined in each class to maintain architectural flexibility.
  
### GridUpdateCallback

When training **KAN-based models**, the grid can be **iteratively refined** during training for better performance.

To support this, logic from the original [`[pykan](https://github.com/KindXiaoming/pykan)`](https://github.com/KindXiaoming/pykan) implementation has been adapted to define a **custom callback** named `GridUpdateCallback`. This callback automatically updates the grid at specified training steps, improving model accuracy and convergence.

This callback has been **tested successfully** and demonstrates **improved results** in practice.

An example usage is provided in:
`examples/nbeats_with_kan.py`

### Checklist

- [x] Linked issues (if existing)
- [x] Used pre-commit hooks when committing to ensure that code is compliant with hooks. Install hooks with `pre-commit install`.
      To run hooks independent of commit, execute `pre-commit run --all-files`

Make sure to have fun coding!

Sohaib-Ahmed21 commented:
@fkiraly @benHeid can you kindly review/merge it so I integrate **NBEATSX** modification in NBEATS without conflicts as I have asked @julian-fong and he is not working on NBEATSX.

benHeid commented on pytorch_forecasting/models/nbeats/_nbeats.py:
```diff
@@ -30,6 +30,19 @@ def __init__(
+        scale_base_mu: float = 0.0,
+        scale_base_sigma: float = 1.0,
+        scale_sp: float = 1.0,
+        base_fun: callable = torch.nn.SiLU(),
```
Please avoid all mutable objects as defaults. Since changing these mutable objects might lead to side-effects that are difficult to debug.

benHeid commented on pytorch_forecasting/models/nbeats/_nbeats.py:
```diff
@@ -30,6 +30,19 @@ def __init__(
+        scale_sp: float = 1.0,
+        base_fun: callable = torch.nn.SiLU(),
+        grid_eps: float = 0.02,
+        grid_range: List[int] = [-1, 1],
```
Please avoid all mutable objects as defaults. Since changing these mutable objects might lead to side-effects that are difficult to debug.

benHeid commented on pytorch_forecasting/models/nbeats/_nbeats.py:
```diff
@@ -47,48 +60,86 @@ def __init__(
-        including ensembles of traditional statical methods in the M4 competition. The M4 competition is arguably
-        the most
-        important benchmark for univariate time series forecasting.
+        forecasting <http://arxiv.org/abs/1905.10437>`_. The network has (if
```
Why are that many docstrings changed? Was this a linter?

Sohaib-Ahmed21 replied:
yes before lines in docstring had >88 characters but linter didn't allow this much length before committing this time.

benHeid commented on pytorch_forecasting/models/nbeats/_nbeats.py:
```diff
@@ -47,48 +60,86 @@ def __init__(
-            backcast_loss_ratio: weight of backcast in comparison to forecast when calculating the loss.
-                A weight of 1.0 means that forecast and backcast loss is weighted the same (regardless of backcast and
-                forecast lengths). Defaults to 0.0, i.e. no weight.
+            num_grids :  Parameter for KAN layer. the number of grid intervals = G.
```
Unsure I have overseen the description of `use_kan`. Please check.

Sohaib-Ahmed21 replied:
Oh I forgot to add, will add it.

benHeid commented on pytorch_forecasting/models/nbeats/_nbeats.py:
```diff
@@ -47,48 +60,86 @@ def __init__(
-            backcast_loss_ratio: weight of backcast in comparison to forecast when calculating the loss.
-                A weight of 1.0 means that forecast and backcast loss is weighted the same (regardless of backcast and
-                forecast lengths). Defaults to 0.0, i.e. no weight.
+            num_grids :  Parameter for KAN layer. the number of grid intervals = G.
```
For all parameters that are only used if use_kan == True, please add a statement that these are ignored if kan is not set to True.

benHeid commented on pytorch_forecasting/models/nbeats/kan_layer.py:
```diff
@@ -0,0 +1,528 @@
+import numpy as np
```
The license of the original implementation is MIT. So in theory it is okay to copy the file. However, please add some credits at the top of the file.

Alternatively, we could think about adding KAN as a dependency. 

@fkiraly do you have any additions on that matter?

Sohaib-Ahmed21 replied:
Okay, I will add the appropriate credits at the top of the file. Additionally, I agree that adding KAN as a dependency—perhaps as a soft dependency—seems like a good idea, especially considering its increasing relevance in time series forecasting.

benHeid replied:
@fkiraly pinging you again to check if this is okay for you :)

fkiraly replied:
which package are we exactly planning to add as a soft dep?

If it is a single layer, I think copying it over and including the license is perhaps better for now, because we do not have machinery to manage soft dependencies (like `scikit-base` or similar).

The proposed design in here https://github.com/sktime/enhancement-proposals/pull/39 would allow that, but right now I think this would require a significant amounts of custom code to handle.

Or, is there an easy way that I am not seeing how the soft dependency import would work for part of the NN?

Sohaib-Ahmed21 replied:
> which package are we exactly planning to add as a soft dep?
> 
It is `pykan` library , reference https://pypi.org/project/pykan/

> If it is a single layer, I think copying it over and including the license is perhaps better for now, because we do not have machinery to manage soft dependencies (like `scikit-base` or similar).
> 
> The proposed design in here [sktime/enhancement-proposals#39](https://github.com/sktime/enhancement-proposals/pull/39) would allow that, but right now I think this would require a significant amounts of custom code to handle.
> 
> Or, is there an easy way that I am not seeing how the soft dependency import would work for part of the NN?

yes it is a single layer, only used in `NBEATS`. Also the library `pykan` has much more, but we only need this. I have implemented what you already suggested above. I have copied it over and included the license.

fkiraly replied:
ok, makes sense, as long as the license points to the original source and is provided in full form (assuming it hsa the usual reqiurement to reproduce)

benHeid reviewed (changes requested):
Not fully reviewed yet. Will continue in the next days. But I share my current comments so that you already receive some feedbacks. 

When training KANs, the grid can be iteratively be refined. I wonder, if there is a way to implement this also here. However, this might probably more difficult and require changes to the trainer. So probably out of scope for this PR. Do you have opinions on that?

Sohaib-Ahmed21 commented:
> Not fully reviewed yet. Will continue in the next days. But I share my current comments so that you already receive some feedbacks.

Thanks! Will address these reviews soon.

Sohaib-Ahmed21 commented:
> When training KANs, the grid can be iteratively be refined. I wonder, if there is a way to implement this also here. However, this might probably more difficult and require changes to the trainer. So probably out of scope for this PR. Do you have opinions on that?

I'll explore this and share my thoughts.

Sohaib-Ahmed21 commented:
@benHeid I have addressed the reviews. Kindly review the updated PR.

> When training KANs, the grid can be iteratively be refined. I wonder, if there is a way to implement this also here. However, this > might probably more difficult and require changes to the trainer. So probably out of scope for this PR. Do you have opinions on > that?

To address this, I have taken logic from original implementation of **pykan** library and made custom Callback i.e. `GridUpdateCallback` which updates grid after specified steps. This has been tested and working fine. Your thoughts will be helpful in this regard.

Sohaib-Ahmed21 commented:
@benHeid @fkiraly  I have addressed the reviews.  Kindly review it so I proceed integrating NBEATSX without conflicts.

benHeid commented on pytorch_forecasting/models/nbeats/grid_callback.py:
I would appreciate if you can add an example on how to use it. Either in the docstring or in a notebook. Otherwise it is implemented but unclear on how to use it.

benHeid commented on pytorch_forecasting/models/nbeats/kan_layer.py:
```diff
@@ -0,0 +1,367 @@
+                grid_adaptive[:, [0]]
+                - margin
+                + h
+                * torch.arange(
```
Ensure that the output of torch.arange is on the same device as h and margin. E.g., if everything is on cuda or mps, this will fail since the default is that the tensor from torch.arange is on cpu.

benHeid reviewed (commented):
I only have one last comment. But I also would like to hear @fkiraly opinion on the license of the kan_layer

fkiraly reviewed (changes requested):
I do have one major request though: This PR seems to be changing the parametrization of `NBEATS` in major ways, and it essentially alters the structure.

May I hence request to, instead of modifying the existing class, to add a new class?
There may be common code that could be refactored, I just think that it is better to keep the class `NBEATS` - a common method - as a conceptual reference to the original NBEATS algorithm.

Variants and modifications I would provide as new classes - there are many modifications of NBEATS; so it seems like a suboptimal design decision to merge one of those (albeit a popular one) in the reference implementation.

Sohaib-Ahmed21 commented:
> I do have one major request though: This PR seems to be changing the parametrization of `NBEATS` in major ways, and it essentially alters the structure.
> 
> May I hence request to, instead of modifying the existing class, to add a new class? There may be common code that could be refactored, I just think that it is better to keep the class `NBEATS` - a common method - as a conceptual reference to the original NBEATS algorithm.
> 
> Variants and modifications I would provide as new classes - there are many modifications of NBEATS; so it seems like a suboptimal design decision to merge one of those (albeit a popular one) in the reference implementation.

Sounds good, thanks for the review! I will make separate class for NBEATS using KAN keeping the original algorithm separate and will update the PR.

fkiraly commented:
FYI @benHeid, thoughts/opinions?

benHeid commented:
I see your point. However, I fear that we will introduce a lot of duplicated code. Thus, we might think about refactoring strategies.

But I am also happy with a new class.

Sohaib-Ahmed21 commented:
> I see your point. However, I fear that we will introduce a lot of duplicated code. Thus, we might think about refactoring strategies.
> 
> But I am also happy with a new class, since this reduces also the complexity.

I'll try to reuse as much code as possible like submodules, etc. So should I move towards new class implementation?

Sohaib-Ahmed21 commented:
@fkiraly @benHeid kindly review the udpated changes.  Thanks!

fkiraly reviewed (changes requested):
LGTM - could we make sure that the new modules in `nbeats` are private? That gives us more flexibility for downwards compatible changes, later.

Sohaib-Ahmed21 commented:
> LGTM - could we make sure that the new modules in `nbeats` are private? That gives us more flexibility for downwards compatible changes, later.

Thanks for the review, I have made the changes and I think it's ready for merge now.

Sohaib-Ahmed21 commented:
@fkiraly this seems ready for merge. Kindly review/merge it. Thanks!

Sohaib-Ahmed21 commented:
The failing test seems unrelated to PR. Also, all reviews have been addressed here. Kindly review/merge it. Thanks!

Sohaib-Ahmed21 commented:
This PR seems ready for merge. @fkiraly @benHeid kindly review/merge it. Thanks!

benHeid commented on pytorch_forecasting/models/nbeats/sub_modules.py:
```diff
@@ -84,7 +172,50 @@ def __init__(
+            min_period: The minimum period used for seasonal patterns. Default: 1.
+            dropout: The dropout rate applied to the fully connected mlp layers to
+                prevent overfitting. Default: 0.1.
+            kan_params (dict): Parameters specific to the KAN layer
```
See comment

benHeid reviewed (changes requested):
Sorry for the late response. 

I guess in the copied modules there are some issues mainly regarding documentation. And I am also unsure if we are using all of the options in the code. 

A big plus would be if you could comment the code that you copied a bit. Since it is hard to understand

Sohaib-Ahmed21 commented:
@benHeid thanks for the detailed review, I'll address it as soon as I get time and update the PR.

Sohaib-Ahmed21 commented on pytorch_forecasting/models/nbeats/_nbeats.py:
```diff
@@ -139,296 +147,9 @@ def __init__(
 
                 self.net_blocks.append(net_block)
-
-    def forward(self, x: Dict[str, torch.Tensor]) -> Dict[str, torch.Tensor]:
```
Yes, it has been copied there.

Sohaib-Ahmed21 commented on pytorch_forecasting/models/nbeats/_nbeats_adapter.py:
The code is copied from [_nbeats.py](https://github.com/sktime/pytorch-forecasting/blob/main/pytorch_forecasting/models/nbeats/_nbeats.py). Only the `__init__` method is modified, as it's customized in child classes depending on whether KAN variant is used or not.

Sohaib-Ahmed21 commented on pytorch_forecasting/models/nbeats/_kan_layer.py:
```diff
@@ -0,0 +1,364 @@
+            shape (batch, in_dim, G+k). G: the number of grid intervals,
+            k: spline order.
+
+    Example
```
Enhanced example with detailed comments on how to run and tested it.

Sohaib-Ahmed21 commented on pytorch_forecasting/models/nbeats/_kan_layer.py:
```diff
@@ -0,0 +1,364 @@
+
+        Example
+        -------
+        >>> from kan.KANLayer import *
```
Enhanced example with detailed comments on how to run and tested it.

Sohaib-Ahmed21 commented on pytorch_forecasting/models/nbeats/_kan_layer.py:
```diff
@@ -0,0 +1,364 @@
+        def get_grid(num_interval):
+            ids = [int(batch / num_interval * i) for i in range(num_interval)] + [-1]
+            grid_adaptive = x_pos[ids, :].permute(1, 0)
+            margin = 0.00
```
You're right, removed it!

Sohaib-Ahmed21 commented on pytorch_forecasting/models/nbeats/_kan_layer.py:
```diff
@@ -0,0 +1,364 @@
+        y = torch.sum(y, dim=1)
+        return y
+
+    def update_grid_from_samples(self, x, mode="sample"):
```
Yes, this was used with default params, removed the unused code!

Sohaib-Ahmed21 commented:
@benHeid I have addressed the reviews. Kindly review the updated PR,  thanks!

fkiraly reviewed (changes requested):
Nice!

Could you kindly also add some explanation in the PR fist post about the changes to the structure, and what goes in the common base class? What is going on with the grid callback?

Also requesting review by @phoeenniixx and @PranavBhatP

Optionally, could you check in the `layers` module whether we can use that to avoid duplicate layer specifications?

phoeenniixx commented on pytorch_forecasting/models/nbeats/_kan_layer.py:
```diff
@@ -0,0 +1,413 @@
+    ):
+        """'
+        Initialize a KANLayer
+
```
Can you move this docstring above just after the class definition?

phoeenniixx commented on pytorch_forecasting/models/nbeats/_nbeatskan.py:
```diff
@@ -0,0 +1,235 @@
+        shown to consistently outperform N-BEATS.
+
+        Args:
+            stack_types: One of the following values: “generic”, “seasonality" or
```
Please use numpydoc style doctsring

phoeenniixx commented on pytorch_forecasting/models/nbeats/_kan_layer.py:
```diff
@@ -0,0 +1,413 @@
+    Then use:
+
+    >>> from kan.spline import B_batch
+    >>> import torch
```
Can we have an example for `pytorch_forecasting`'s implementation as I think it will be helpful for users using this implementation :)

phoeenniixx replied:
I think then if you are adding egs, some egs how the `ptf` implementation can be used will also be helpful, not necessarily here, maybe in `NBeatsKAN`? Its just a suggestion -non blocking

Sohaib-Ahmed21 replied:
Added `examples` folder's example link in docstring of `NBEATSKAN`.

phoeenniixx commented on pytorch_forecasting/models/nbeats/_nbeatskan.py:
```diff
@@ -0,0 +1,235 @@
+        forecasting <http://arxiv.org/abs/1905.10437>`_. The network has (if
+        used as ensemble) outperformed all other methods including ensembles of
+        traditional statical methods in the M4 competition. The M4 competition is
+        arguably the most important benchmark for univariate time series forecasting.
```
I think it would be good to mention the paper for `KAN` blocks here and how it works (just an overview) and how it is different from the original `NBeats`

phoeenniixx commented on pytorch_forecasting/models/nbeats/_nbeatskan_pkg.py:
```diff
@@ -0,0 +1,39 @@
+    @classmethod
+    def get_test_train_params(cls):
+        """Return testing parameter settings for the trainer."""
+        return [{"backcast_loss_ratio": 1.0}]
```
Can we have some more params? Just to see any edge cases or anything?

Sohaib-Ahmed21 commented:
> Could you kindly also add some explanation in the PR fist post about the changes to the structure, and what goes in the common base class? What is going on with the grid callback?

Updated the first post to include explanations on the structure changes, common base class, and grid callback as requested.

> Optionally, could you check in the layers module whether we can use that to avoid duplicate layer specifications?

I checked it and currently blocks are reused from `sub_modules.py` and are not duplicated.

Sohaib-Ahmed21 commented:
@fkiraly @benHeid kindly review/merge this pull request, thanks!

phoeenniixx commented:
> > Optionally, could you check in the layers module whether we can use that to avoid duplicate layer specifications?
> 
> I checked it and currently blocks are reused from `sub_modules.py` and are not duplicated.

I think we can also move some of the layers (like `KAN` layer) to the `layers` right? I think these layers can be reused by other models as well? 

Also, a question for @fkiraly: Imo, we could also move networks present in `sub_modules` to `layers`, or is it better to keep them here along with `NBeats` only, as other models might not use them?

Sohaib-Ahmed21 commented:
> I think we can also move some of the layers (like `KAN` layer) to the `layers` right? I think these layers can be reused by other models as well?

Yes, in the future,  `KANLayer` might be reused...I think then [`GridUpdateCallback`](https://github.com/Sohaib-Ahmed21/pytorch-forecasting/blob/5a31a583ab8108d44843e83093e852d81511e516/pytorch_forecasting/models/nbeats/_grid_callback.py#L4) should also be moved to `layers`. Thoughts?

phoeenniixx commented:
> > I think we can also move some of the layers (like `KAN` layer) to the `layers` right? I think these layers can be reused by other models as well?
> 
> Yes, in the future, `KANLayer` might be reused...I think then [`GridUpdateCallback`](https://github.com/Sohaib-Ahmed21/pytorch-forecasting/blob/5a31a583ab8108d44843e83093e852d81511e516/pytorch_forecasting/models/nbeats/_grid_callback.py#L4) should also be moved to `layers`. Thoughts?

Yes this can also be added!
I think we need to decide which parts of this PR are reusable and can be moved to `layers`

fkiraly commented:
> Also, a question for @fkiraly: Imo, we could also move networks present in `sub_modules` to `layers`, or is it better to keep them here along with `NBeats` only, as other models might not use them?

I think ultimately all layers should move to `layers`, although moving layers not changed or affected by this PR is out of scope. But I would move new layers, or layers that now get imported by both nbeats variants to there.

fkiraly reviewed (changes requested):
@Sohaib-Ahmed21, not sure if you saw my last message - could you kindly let us know if you are still working on this? If so:

* please mofe layers to the `layers` folder
* add a `pkg` for the new KAN based NBEATS so it is automatically collected by the test framework and tested

Sohaib-Ahmed21 commented:
> @Sohaib-Ahmed21, not sure if you saw my last message - could you kindly let us know if you are still working on this? If so:
> 
> * please move layers to the `layers` folder
> * add a `pkg` for the new KAN based NBEATS so it is automatically collected by the test framework and tested

Hi @fkiraly, thanks for the catch-up. I'm still on it, will do by this weekend. Got busy with the job lately.

fkiraly reviewed (changes requested):
the tests are running now and this looks like a genuine failure in the new code

Sohaib-Ahmed21 commented:
> the tests are running now and this looks like a genuine failure in the new code

Okay, I'll look into and fix it.

fkiraly commented:
ah, seems to work now! Can you explain why it was failing?

Sohaib-Ahmed21 commented:
> ah, seems to work now! Can you explain why it was failing?

It failed because the test data used before could give `TweedieLoss` invalid targets; now the dataloader ensures valid non-negative targets, so the loss runs without errors.

Sohaib-Ahmed21 commented:
@fkiraly kindly review/merge this pull request, thanks!

fkiraly commented on pytorch_forecasting/models/nbeats/__init__.py:
```diff
@@ -1,17 +1,17 @@
 
 __all__ = [
     "NBeats",
-    "NBEATSGenericBlock",
```
to avoid breaking existing code, we need to keep exporting these in the current location, so import them from the networks location.

We should also add a `# todo v2: remove imports` to those, at a major we can break.

fkiraly commented on pytorch_forecasting/models/nbeats/sub_modules.py:
```diff
@@ -1,219 +0,0 @@
-    return b_ls, f_ls
-
-
-class NBEATSBlock(nn.Module):
```
same comment here - I would also import the blocks here

fkiraly reviewed (changes requested):
Looks good to me!

Very small thing, I think we should restore the exports of the layers, to avoid breaking user code that imported these previously public classes.

fkiraly reviewed (approved):
Looks good now, great addition!
